### PR TITLE
Resolve path as relative path in sandbox rather than using absolute path

### LINKED ISSUE #23: relative path doesn't work with dune 3.0
If we change `(lang dune 1.0)` to `(lang dune 3.0)` (https://github.com/johnwhitington/ppx_blob/blob/0.7.2/dune-project#L1
the build will failed with following error.
```
$ make all
dune build @install @example
File "example/quine.ml", line 1, characters 13-31:
1 | print_string [%blob "quine.ml"]
                 ^^^^^^^^^^^^^^^^^^
Error: [%blob] could not find or load file quine.ml
```

I found that [`Ocaml_common.Location.absolute_path path`](https://github.com/johnwhitington/ppx_blob/blob/0.7.2/src/ppx_blob.ml#L9) no longer return a real path but prefix the argument with `/workspace_root/`.
Since there is no `/workspace_root/example/quine.ml`, a `Not_found` is thrown.

### PR DESCRIPTION
**Pull request by actionshrimp**
We don't seem to need to use `absolute_path` if the relevant `preprocessor_deps` stanzas are added - in this case it looks like we can just load the file as a relative path from the dune sandbox.

This seems to fix #23 when I've tested it with dune, although I may have missed some subtleties here or broken other non-dune paths if those are supported.

I had to update the `dune-project` `lang` for ppx_blob itself in order to replicate the issue in the tests.

### CODE REVIEW

File: test_file
Comment (sim642) [inline test_file]: It might be better to make the test directory deeper such that this test file doesn't end up at the root of the project.

**Review by sim642 — APPROVED** [review]
I tested this on a project of mine and it indeed fixes the problem.

File: test_file
Comment (actionshrimp) [reply test_file]: Yep, good call - have tidied that up a bit in the next commit.

**Comment by actionshrimp** [comment]
Hi, would it be possible for someone with access to merge this when they get a chance? Thanks!

**Comment by sim642** [comment]
By the way, since ppx_blob is very small, it's no problem to just vendor the fixed version into some dune project which requires it. I've done it here: https://github.com/goblint/analyzer/tree/master/src/vendor/ppx_blob.
That way there's no need to wait for a fixed opam release.

**Comment by Niols** [comment]
Hello. Any way to see this merged and pushed into opam-repository? In the meantime, I did go for vendoring, thank you @sim642!

**Comment by johnwhitington** [comment]
No one has answered the question of whether this is a dune-specific workaround, or a genuine bug fix. And could it negatively affect non-Dune workflows?

Dune is popular, and deservedly so, but it isn't OCaml.

**Comment by Niols** [comment]
> Dune is popular, and deservedly so, but it isn't OCaml.

Wholeheartedly agree. There should probably be tests for various workflows (eg. Dune, manual, OCamlBuild, etc.) for this reason.

**Comment by bbc2** [comment]
If that's OK with you I will look into adding some integration tests to ppx_blob. That way we will understand better what we might break.

**Comment by bbc2** [comment]
I'm done with the tests. You can see the results for the `master` branch here: https://github.com/johnwhitington/ppx_blob/pull/25. The impact of this PR on these tests would be (for now): https://github.com/johnwhitington/ppx_blob/commit/c7d89cd2d5168e200b5a874e0b4579a291d650ca.

During my investigation, the few issues I found in ppx_blob were not related to this PR:

- OCamlbuild
- Absolute paths (e.g. `/etc/hostname` in my example).
- Unreliable working directory for PPX invocation (which makes workspace-relative directories hard to use with OCaml-LSP). This might be related to https://github.com/ocaml/dune/issues/4479.
  - This would be an argument in favor of the use of source-file-relative paths, which the present PR enables for Dune lang 3+.

I'm therefore mostly good with this fix. I would just avoid adding a dependency on Dune 3+ we can keep fixing ppx_blob for a wide audience. In the future we might require Dune 3+ but there's doesn't seem to be a need now if we have good integration tests.

What do you think?

**Comment by johnwhitington** [comment]
@bbc2 thanks for this detailed work. I leave it entirely up to you - merge when you're happy.

I haven't understood ppx_blob since the afternoon I wrote it!

Actually, an argument could be made that this is the left-pad of PPXs and should really just be in the compiler, with the paths passed on the ocamlc/ocamlopt command line.

**Comment by bbc2** [comment]
I'll be merging this, thank you for your contribution!

For information, I pushed an updated commit:

- The code of ppx_blob and the tests is the same as in the initial state of the PR.
- `(lang dune 3)` was changed to `(lang dune 1.11)` to avoid adding dependency on Dune 3 just now.
  - This is OK because thanks to #25 we have integration tests which check the behavior of ppx_blob with `(lang dune 3)` (among other things).

**Pull request merged**
